features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin_new/source/src/app/components/Throttling/Custom/AddEdit.jsx

```diff
@@ -0,0 +1,337 @@
+
+        return errorText;
+    };
+
+    const formSaveCallback = () => {
+        const formErrors = getAllFormErrors();
+        if (formErrors !== '' && formErrors !== 'false') {
+            Alert.error(formErrors);
+            return (false);
+        }
+
+        let promisedAddCustomPolicy;
+        if (state.siddhiQuery.indexOf('\n') !== -1) {
+            const siddhiQueryValue = state.siddhiQuery.split('\n').join(' ');
+            delete (state.siddhiQuery);
+            state.siddhiQuery = siddhiQueryValue;
+        }
+        const customPolicy = state;
+
+        if (dataRow) {
+            const { policyId } = dataRow;
+            promisedAddCustomPolicy = restApi.updateCustomPolicy(policyId,
+                customPolicy);
+            promisedAddCustomPolicy = new Promise((resolve, reject) => {
+                promisedAddCustomPolicy
+                    .then(() => {
+                        resolve(
+                            <FormattedMessage
+                                id='Throttling.Application.Policy.policy.edit.success'
+                                defaultMessage='Application Rate Limiting Policy edited successfully.'
+                            />,
+                        );
+                    })
+                    .catch((error) => {
+                        const { response } = error;
+                        if (response.body) {
+                            reject(response.body.description);
+                        }
+                    })
+                    .finally(() => {
+                        updateList();
+                    });
+            });
+        } else {
+            promisedAddCustomPolicy = restApi.addCustomPolicy(
+                customPolicy,
+            );
+            promisedAddCustomPolicy = new Promise((resolve, reject) => {
+                promisedAddCustomPolicy
+                    .then(() => {
+                        resolve(
+                            <FormattedMessage
+                                id='Throttling.Application.Policy.policy.add.success'
+                                defaultMessage='Custom Policy added successfully.'
+                            />,
+                        );
+                    })
+                    .catch((error) => {
+                        const { response } = error;
+                        if (response.body) {
+                            reject(response.body.description);
+                        }
+                    })
+                    .finally(() => {
+                        updateList();
+                    });
+            });
+        }
+        return (promisedAddCustomPolicy);
+    };
+
+    const dialogOpenCallback = () => {
+        if (dataRow) {
+            setIsEditMode(true);
+            const { policyId } = dataRow;
+            restApi.customPolicyGet(policyId).then((result) => {
+                const formattedSiddhiQuery = sqlFormatter.format(result.body.siddhiQuery);
```

Comment:
Since siddhi is not 100% equal to SQL, this formatter could fail with some siddhi queries don't we need to handle such situations ?

Reply:
//TODO: Switch to ACE editor and give the siddhi g4 grammar to antlr plugin.